app/project_type/git.py

```diff
@@ -144,45 +136,63 @@ def _fetch_project(self):
-        self._git_remote_command_executor.execute(fetch_command, cwd=self._repo_directory)
+        self._execute_git_command_in_repo_and_raise_on_failure(
+            'fetch --update-head-ok {} {}'.format(self._remote, self._branch),
+            'Could not fetch specified branch "{}" from remote "{}".'.format(self._branch, self._remote)
+        )
 
         # Validate and convert the user-specified hash/refspec to a full git hash
-        self._hash = self._execute_in_repo_and_raise_on_failure(
-            'git rev-parse {}'.format(self._hash),
+        self._hash = self._execute_git_command_in_repo_and_raise_on_failure(
+            'rev-parse {}'.format(self._hash),
             'Could not rev-parse "{}" to a commit hash.'.format(self._hash)
         ).strip()
 
         # Save this hash as a local ref. Named local refs are necessary for slaves to fetch correctly from the master.
         # The local ref will be passed on to slaves instead of the user-specified branch.
         self._local_ref = 'refs/clusterrunner/' + self._hash
-        update_ref_command = 'git update-ref {} {}'.format(self._local_ref, self._hash)
-        self._execute_in_repo_and_raise_on_failure(update_ref_command, 'Could not update local ref.')
+        update_ref_command = 'update-ref {} {}'.format(self._local_ref, self._hash)
+        self._execute_git_command_in_repo_and_raise_on_failure(update_ref_command, 'Could not update local ref.')
 
         # The '--' option acts as a delimiter to differentiate values that can be "tree-ish" or a "path"
-        reset_command = 'git reset --hard {} --'.format(self._hash)
-        self._execute_in_repo_and_raise_on_failure(reset_command, 'Could not reset Git repo.')
+        reset_command = 'reset --hard {} --'.format(self._hash)
+        self._execute_git_command_in_repo_and_raise_on_failure(reset_command, 'Could not reset Git repo.')
 
-        self._execute_in_repo_and_raise_on_failure('git clean -dfx', 'Could not clean Git repo.')
+        self._execute_git_command_in_repo_and_raise_on_failure('clean -dfx', 'Could not clean Git repo.')
 
-    def _execute_in_repo_and_raise_on_failure(self, command, message):
+    def _execute_git_command_in_repo_and_raise_on_failure(self, command, message='Error executing git command.'):
         """
+        :param command: The git command to execute, e.g., "fetch origin"
+        :type command: string
+        :param message: The human readable error message to log if the command fails
+        :type message: string
+        :return: The output of the process (stdout and stderr)
         :rtype: string
         """
-        return self._execute_and_raise_on_failure(command, message, self._repo_directory)
+        strict_host_key_setting = 'yes' if Configuration['git_strict_host_key_checking'] else 'no'
+        git_ssh_args = '-o BatchMode=yes -o StrictHostKeyChecking={}'.format(strict_host_key_setting)
```

Comment:
Actually, from manual testing it looks like `BatchMode=yes` will still prevent a prompt even in the case where there is a remote host key mismatch. If `StrictHostKeyChecking=yes` then the ssh command will fail immediately, and if `StrictHostKeyChecking=no` then ssh will succeed despite the key mismatch (although it will print out a warning).

Given that behavior, I lean towards not including the `UserKnownHostsFile=/dev/null` option. I think the only effect of including it is that is suppresses the mismatched hosts warning that gets printed to stdout during git operations. It doesn't do anything to prevent a prompt.